src/main/scala/com/daml/projection/javadsl/BatchSource.scala

```diff
@@ -112,4 +134,57 @@ object BatchSource {
           J.TreeEvent.fromProtoTreeEvent(ST.TreeEvent.toJavaProto(event))))
     }
   }.toJava
+
+  @FunctionalInterface
+  trait GetContractTypeId[E] {
+    def fromEvent(event: E): ju.Optional[J.Identifier]
```

Comment:
I'm now preferring to call it from
```suggestion
    def from(e: E): ju.Optional[J.Identifier]
```
Since it is not necessarily an event, it can be anything (a contract for instance)